src/main/python/ttconv/isd.py

```diff
@@ -0,0 +1,346 @@
+
+      return (begin_time, end_time)
+
+    def sig_times(element: model.ContentElement, parent_begin: Fraction, parent_end: typing.Optional[Fraction]):
+
+      # add signficant times for the element
+
+      begin_time, end_time = make_absolute(element.get_begin(), element.get_end(), parent_begin, parent_end)
+
+      s_times.add(begin_time)
+
+      if end_time is not None:
+        s_times.add(end_time)
+
+      # add signficant times for any animation step 
+
+      for anim_step in element.iter_animation_steps():
+        anim_begin_time, anim_end_time = make_absolute(anim_step.begin, anim_step.end, parent_begin, parent_end)
+
+        s_times.add(anim_begin_time)
+
+        if anim_end_time is not None:
+          s_times.add(anim_end_time)
+
+      # add signficant times for the children of the element 
+
+      for child_element in iter(element):
+        sig_times(child_element, begin_time, end_time)
+
+    s_times = set()
+
+    # add significant times for regions
+
+    for region in doc.iter_regions():
+      sig_times(region, 0, None)
+
+    # add significant times for body and its descendents
+
+    sig_times(doc.get_body(), 0, None) 
+
+    return s_times
+
+  @staticmethod
+  def from_model(doc: model.Document, offset: Fraction) -> typing.Optional[ISD]:
+    '''Creates an ISD from a snapshot of a Document `doc` at a given time offset `offset`
+    '''
+    isd = ISD(doc)
+
+    for region in doc.iter_regions():
+      root_region = ISD.process_element(isd, offset, region, None, None, region)
+
+      if root_region is not None:
+        isd.put_region(root_region)
+
+    return isd
+
+
+  @staticmethod
+  def compute_style(_style_property: styles.StyleProperty, value: typing.Any) -> typing.Any:
+    return value
+
+  @staticmethod
+  def process_element(isd: ISD,
+                      offset: Fraction,
+                      selected_region: model.Region,
+                      inherited_region: typing.Optional[model.Region],
+                      parent: model.ContentElement,
+                      element: model.ContentElement
+                      ) -> typing.Optional[model.ContentElement]:
+
+    doc = element.get_doc()
+
+    # convert offset to local temporal coordinates
+
+    if not isinstance(element, (model.Region, model.Body)):
+      offset = offset - (parent.get_begin() or 0)
+
+    # prune if temporally inactive
+    
+    if element.get_begin() is not None and element.get_begin() > offset:
+      return None
+
+    if element.get_end() is not None and element.get_end() > offset:
+      return None
+
+    # associated region is that associated with the element, or inherited otherwise
+
+    associated_region = element.get_region() or inherited_region
+
+    # prune the element if either:
+    # * the element has children and the associated region is neither the default nor the root region
+    # * the element has no children and the associated region is not the root region
+
+    if (
```

Comment:
Are these parentheses useful?

Reply:
They are necessary for multiline logical expressions AFAIK.